android/src/main/java/com/horcrux/svg/RenderableView.java

```diff
@@ -329,71 +330,65 @@ public void setPropList(@Nullable ReadableArray propList) {
     invalidate();
   }
 
-  private static double saturate(double v) {
-    return v <= 0 ? 0 : (v >= 1 ? 1 : v);
-  }
-
   void render(Canvas canvas, Paint paint, float opacity) {
     MaskView mask = null;
+
     if (mMask != null) {
       SvgView root = getSvgView();
       mask = (MaskView) root.getDefinedMask(mMask);
     }
-    if (mask != null) {
-      Rect clipBounds = canvas.getClipBounds();
-      int height = clipBounds.height();
-      int width = clipBounds.width();
 
-      Bitmap maskBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
-      Bitmap original = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
-      Bitmap result = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
+    if (mask != null) {
+      // https://www.w3.org/TR/SVG11/masking.html
+      // Adding a mask involves several steps
+      // 1. applying luminanceToAlpha to the mask element
+      // 2. merging the alpha channel of the element with the alpha channel from the previous step
+      // 3. applying the result from step 2 to the target element
 
-      Canvas originalCanvas = new Canvas(original);
-      Canvas maskCanvas = new Canvas(maskBitmap);
-      Canvas resultCanvas = new Canvas(result);
+      Paint dstInPaint = new Paint();
+      dstInPaint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.DST_IN));
 
-      // Clip to mask bounds and render the mask
+      // calculate mask bounds
       float maskX = (float) relativeOnWidth(mask.mX);
       float maskY = (float) relativeOnHeight(mask.mY);
       float maskWidth = (float) relativeOnWidth(mask.mW);
       float maskHeight = (float) relativeOnHeight(mask.mH);
-      maskCanvas.clipRect(maskX, maskY, maskWidth + maskX, maskHeight + maskY);
 
-      Paint maskPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
-      mask.draw(maskCanvas, maskPaint, 1);
+      // step 3 - combined layer
```

Comment:
Why is the step 3 the first step? Maybe we should call it `Preparation for step 3` or something like that?